handlers.go

```diff
@@ -4302,3 +4305,49 @@ func (s *Server) handleRunGc(c echo.Context) error {
 
 	return nil
 }
+
+func (s *Server) handleGateway(c echo.Context) error {
+	proto, cc, segs, err := gateway.ParsePath("/" + c.Param("path"))
+	if err != nil {
+		return err
+	}
+
+	redir, err := s.checkGatewayRedirect(proto, cc, segs)
+	if err != nil {
+		return err
+	}
+
+	if redir == "" {
+		s.gwayHandler.ServeHTTP(c.Response().Writer, c.Request())
+		return nil
+	}
+
+	return c.Redirect(302, redir)
+}
+
+func (s *Server) checkGatewayRedirect(proto string, cc cid.Cid, segs []string) (string, error) {
+	if proto != "ipfs" {
+		return fmt.Sprintf("https://ipfs.io/%s/%s/%s", proto, cc, strings.Join(segs, "/")), nil
+	}
+
+	var cont Content
+	if err := s.DB.First(&cont, "cid = ? and active and not offloaded", &util.DbCID{cc}).Error; err != nil {
+		if xerrors.Is(err, gorm.ErrRecordNotFound) {
+			return "", nil
+		}
+		return "", err
+	}
+
+	if cont.Location == "local" {
+		return "", nil
+	}
+
+	var shuttle Shuttle
+	if err := s.DB.First(&shuttle, "handle = ?", cont.Location).Error; err != nil {
+		return "", err
+	}
+
+	// TODO: maybe check if the shuttle is online?
```

Comment:
Yeah, I'd do that, with a timed-memoize, like 30 secs or something, and then bat folks to regular dweb.link.